File: static/utils/vesta.js
/**
 *  This work is based on original code developed and copyrighted by TNO 2020.
 *  Subsequent contributions are licensed to you by the developers of such code and are
 *  made available to the Project under one or several contributor license agreements.
 *
 *  This work is licensed to you under the Apache License, Version 2.0.
 *  You may obtain a copy of the license at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Contributors:
 *      TNO         - Initial implementation
 *  Manager:
 *      TNO
 */

class Vesta {
    constructor() {
        this.restrictions_list = [];
        this.measures_list = [];

        this.current_measures = null;
        this.current_area_id = null;

        this.initSocketIO();
    }

    initSocketIO() {
        console.log("Registering socket io bindings for Vesta plugin")

        socket.on('vesta_restrictions_data', function(vesta_data) {
//            console.log(vesta_data);

            vesta_plugin.current_measures = vesta_data['current_measures'];
            vesta_plugin.current_area_id = vesta_data['area_id'];

            sidebar.setContent(vesta_plugin.create_vesta_sidebar_content(vesta_data).get(0));
            sidebar.show();
            vesta_plugin.load_measures();
        });

        this.get_vesta_restrictions_list();
    }

    get_vesta_restrictions_list() {
        console.log('http get /vesta_restrictions');
        $.ajax({
            url: resource_uri + '/vesta_restrictions',
            success: function(data){
                for (let i=0; i<data.length; i++) {
                    vesta_plugin.restrictions_list.push({id: data[i]["id"], title: data[i]["title"]});
                }
            }
        });
    }

    create_vesta_sidebar_content(data) {
        let $div = $('<div>').attr('id', 'vesta-main-div');
        let $title = $('<h1>').text('VESTA');
        $div.append($title);

        let $select_div = $('<div>').addClass('sidebar-div');
        let $select = $('<select>').attr('id', 'select_restrictions_list');
        this.create_restrictions_list_select($select);
        $select.change(function() { vesta_plugin.load_measures(); });
        $select_div.append($select);
        $div.append($select_div);

        let $measures_div = $('<div>').attr('id', 'measures_div').addClass('sidebar-div');
        $measures_div.append($('<p>').text('Please select your restrictions for ' + data["area_id"]));
        let $measures_p = $('<p>').attr('id', 'measures_p');
        let $measures_button_p = $('<p>').attr('id', 'measures_button_p');
        $measures_div.append($measures_p).append($measures_button_p);
        $div.append($measures_div);

        return $div;
    }

    create_restrictions_list_select($select) {
        for (let i=0; i<this.restrictions_list.length; i++) {
            let $option = $('<option>').attr('value', this.restrictions_list[i]['id']).text(this.restrictions_list[i]['title']);
            $select.append($option);
        }
    }

    load_measures() {
        let $select = $('#select_restrictions_list');
        let selected_value = $select.val();
        console.log(selected_value);

        $.ajax({
            url: resource_uri + '/vesta_restriction/' + selected_value,
            success: function(data){
                if (data.length > 0) {
                    let $ul = $('<ul>').css('list-style-type', 'none');
                    $('#measures_p').empty().append($ul);
                    for (let i=0; i<data.length; i++) {
//                        console.log(data[i]);

                        let $li = $('<li>');
                        let $cb = $('<input>').attr('type', 'checkbox').attr('value', data[i]['id']).attr('name', 'measures');
                        for (let j=0; j<vesta_plugin.current_measures.length; j++) {
                            if (vesta_plugin.current_measures[j]['id'] ==  data[i]['id']) $cb.prop('checked', true);
                        }
                        let $label = $('<label>').attr('for', data[i]["id"]).text(data[i]["name"]);
                        $li.append($cb).append($label);
                        $ul.append($li);
                    }
                    let $button = $('<button>').attr('type', 'button').attr('id', 'id_sel_restr').attr('name', 'Select restrictions').text('Select restrictions');
                    $button.click(function(e){
                        vesta_plugin.set_area_restrictions();
                        sidebar.hide();
                    });
                    $('#measures_button_p').empty().append($button);
                }
            }
        });
    }

    area_restrictions_window(event, id) {
        socket.emit('vesta_area_restrictions', id);
    }

    set_area_restrictions(event) {
        let selected_measures = [];
        $.each($("input[name='measures']:checked"), function(){
            selected_measures.push($(this).val());
        });

        socket.emit('select_area_restrictions', {area_id: this.current_area_id, selected_measures: selected_measures});
    }

    static create(event) {
        if (event.type === 'client_connected') {
            vesta_plugin = new Vesta();
            return vesta_plugin;
        }
        if (event.type === 'add_contextmenu') {
            let layer = event.layer;
            let layer_type = event.layer_type;
            let id = layer.id;
            if (layer_type === 'area') {
                layer.options.contextmenuItems.push({
                    text: 'set VESTA restrictions',
                    icon: resource_uri + 'icons/Vesta.png',
                    callback: function(e) {
                        vesta_plugin.area_restrictions_window(e, id);
                    }
                });
            }
        }
        if (event.type === 'settings_menu_items') {
            let menu_items = {
                'value': 'vesta_plugin_settings',
                'text': 'VESTA plugin',
                'settings_func': function() { return $('<p>').text('VESTA plugin')},
                'sub_menu_items': []
            };

            return menu_items;
        }
    }
}

var vesta_plugin;   // global variable for the Vesta plugin

$(document).ready(function() {
    extensions.push(function(event) { return Vesta.create(event) });
});

File: extensions/vesta.py
#  This work is based on original code developed and copyrighted by TNO 2020.
#  Subsequent contributions are licensed to you by the developers of such code and are
#  made available to the Project under one or several contributor license agreements.
#
#  This work is licensed to you under the Apache License, Version 2.0.
#  You may obtain a copy of the license at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Contributors:
#      TNO         - Initial implementation
#  Manager:
#      TNO

from flask import Flask, jsonify
from flask_socketio import SocketIO, emit
from extensions.settings_storage import SettingsStorage
from extensions.session_manager import get_handler, get_session
from src.process_es_area_bld import get_area_id_from_mapeditor_id
from esdl import esdl
from esdl.processing import ESDLAsset, ESDLEnergySystem
import src.settings as settings
import requests
import json
import uuid
import src.log as log

logger = log.get_logger(__name__)


VESTA_SYSTEM_CONFIG = 'VESTA_CONFIG'


def send_alert(message):
    print(message)
    emit('alert', message, namespace='/esdl')


class Vesta:
    def __init__(self, flask_app: Flask, socket: SocketIO, settings_storage: SettingsStorage):
        self.flask_app = flask_app
        self.socketio = socket
        self.settings_storage = settings_storage
        self.vesta_plugin_settings = self.get_config()
        self.selected_measures = None

        self.register()

    def get_config(self):
        if self.settings_storage.has_system(VESTA_SYSTEM_CONFIG):
            vesta_plugin_settings = self.settings_storage.get_system(VESTA_SYSTEM_CONFIG)
        else:
            vesta_plugin_settings = {
                "edr_url": settings.edr_config['EDR_host'],
                "edr_measures_tags": "measures,vesta"
            }
            self.settings_storage.set_system(VESTA_SYSTEM_CONFIG, vesta_plugin_settings)
        return vesta_plugin_settings

    def register(self):
        logger.info("Registering VESTA extension")

        @self.socketio.on('vesta_area_restrictions', namespace='/esdl')
        def set_area_restrictions(area_id):
            with self.flask_app.app_context():
                esh = get_handler()
                print(area_id)
                area_id = get_area_id_from_mapeditor_id(area_id)

                active_es_id = get_session('active_es_id')
                current_es = esh.get_energy_system(active_es_id)
                top_area = current_es.instance[0].area

                area = ESDLEnergySystem.find_area(top_area, area_id)
                measures = area.measures
                current_area_measures_list = []
                if measures:
                    for m in measures.measure:
                        if isinstance(m, esdl.MeasureReference):
                            this_measure = m.reference
                        else:
                            this_measure = m

                        current_area_measures_list.append({"id": this_measure.id, "name": this_measure.name})

                data = {
                    "current_measures": current_area_measures_list,
                    "area_id": area_id
                }

                emit('vesta_restrictions_data', data)

        @self.socketio.on('select_area_restrictions', namespace='/esdl')
        def select_area_restrictions(measures_data):
            area_id = measures_data['area_id']
            area_id = get_area_id_from_mapeditor_id(area_id)
            measures = measures_data['selected_measures']

            with self.flask_app.app_context():
                esh = get_handler()

                active_es_id = get_session('active_es_id')
                current_es = esh.get_energy_system(active_es_id)
                top_area = current_es.instance[0].area

                area = ESDLEnergySystem.find_area(top_area, area_id)
                print(measures)
                self.set_area_measures(esh, area_id, measures)

        @self.flask_app.route('/vesta_restrictions')
        def get_vesta_restrictions():
            result = self.get_edr_vesta_measures_list()
            if result is None:
                return jsonify([]), 200
            else:
                return jsonify(result), 200

        @self.flask_app.route('/vesta_restriction/<id>')
        def get_vesta_restriction(id):
            result = self.get_edr_vesta_measures(id)
            if result is None:
                return jsonify([]), 200
            else:
                return jsonify(result), 200

    def set_area_measures(self, esh, area_id, measures_list):
        active_es_id = get_session('active_es_id')
        es = esh.get_energy_system(active_es_id)
        top_area = es.instance[0].area

        area_id = get_area_id_from_mapeditor_id(area_id)
        area = ESDLEnergySystem.find_area(top_area, area_id)
        area.measures = None

        if measures_list:
            es_measures = es.measures

            for m_id in measures_list:
                found_measure = False
                if es_measures:
                    for es_m in es_measures.measure:
                        if es_m.id == m_id:
                            m_ref = esdl.MeasureReference()
                            m_ref.id = str(uuid.uuid4())
                            m_ref.reference = es_m
                            found_measure = True
                            if not area.measures:
                                area.measures = esdl.Measures()
                            area.measures.measure.append(m_ref)
                            esh.add_object_to_dict(active_es_id, m_ref)

                if not found_measure:
                    for edr_m in self.selected_measures.measure:
                        if edr_m.id == m_id:
                            # Add measure to es.measures
                            this_m = edr_m.deepcopy()
                            if not es_measures:
                                es_measures = esdl.Measures()
                                es.measures = es_measures
                            es_measures.measure.append(this_m)
                            esh.add_object_to_dict(active_es_id, this_m, True)

                            # Create reference
                            m_ref = esdl.MeasureReference()
                            m_ref.id = str(uuid.uuid4())
                            m_ref.reference = this_m

                            if not area.measures:
                                area.measures = esdl.Measures()
                            area.measures.measure.append(m_ref)
                            esh.add_object_to_dict(active_es_id, m_ref)

    def get_edr_vesta_measures_list(self):
        url = self.vesta_plugin_settings["edr_url"] + "/store/tagged?tag=" + self.vesta_plugin_settings["edr_measures_tags"]

        try:
            r = requests.get(url)
            if r.status_code == 200:
                result = json.loads(r.text)
                return result
            else:
                print('EDR returned unexpected result. Check Vesta plugin settings')
                send_alert('EDR returned unexpected result. Check Vesta plugin settings')
                return None

        except Exception as e:
            print('Error accessing EDR API: ' + str(e))
            send_alert('Error accessing EDR API: ' + str(e))
            return None

    def get_edr_vesta_measures(self, measures_id):
        url = self.vesta_plugin_settings["edr_url"] + "/store/esdl/" + measures_id + "?format=xml"

        try:
            r = requests.get(url)
            if r.status_code == 200:
                result = []
                self.selected_measures = ESDLAsset.load_asset_from_string(r.text)
                for m in self.selected_measures.measure:
                    result.append({
                        "id": m.id,
                        "name": m.name
                    })
                return result
            else:
                print('EDR returned unexpected result. Check Vesta plugin settings')
                send_alert('EDR returned unexpected result. Check Vesta plugin settings')
                return None

        except Exception as e:
            print('Error accessing EDR API: ' + str(e))
            send_alert('Error accessing EDR API: ' + str(e))
            return None
